src/random.c

```diff
@@ -4,12 +4,102 @@
 #include <alloca.h>
 #endif
 
-EWRAM_DATA static u8 sUnknown = 0;
-EWRAM_DATA static u32 sRandCount = 0;
-
 // IWRAM common
-u32 gRngValue;
-u32 gRng2Value;
+rng_value_t gRngValue;
+rng_value_t gRng2Value;
+
+#if HQ_RANDOM == TRUE
+EWRAM_DATA static volatile bool8 sRngLoopUnlocked;
+
+// Do not change STREAM1 without changing Random32().
+#define STREAM1 1
+#define STREAM2 29
+
+// A variant of SFC32 that lets you change the stream.
+// stream cannot be even, but any other (8-bit) value should be OK.
+static inline u32 _SFC32_Next_Stream(struct Sfc32State *state, const u8 stream)
+{
+    const u32 result = state->a + state->b + state->ctr;
+    state->ctr += stream;
+    state->a = state->b ^ (state->b >> 9);
+    state->b = state->c * 9;
+    state->c = result + ((state->c << 21) | (state->c >> 11));
+    return result;
+}
+
+static void SFC32_Seed(struct Sfc32State *state, u16 seed, u8 stream)
+{
+    u32 i;
+    state->a = state->b = 0;
+    state->c = seed;
+    state->ctr = stream;
+    for(i = 0; i < 16; i++)
+    {
+        _SFC32_Next_Stream(state, stream);
+    }
+}
+
+u32 NAKED Random32(void)
+{
+    asm(".thumb\n\
+    push {r4, r5, r6}\n\
+    mov r6, #11\n\
+    ldr r5, =gRngValue\n\
+    ldmia r5!, {r1, r2, r3, r4}\n\
+    @ e (result) = a + b + d++\n\
+    add r1, r1, r2\n\
+    add r0, r1, r4\n\
+    @the argument is the same as STREAM1\n\
+    add r4, r4, #1\n\
```

Comment:
Could you keep this in sync with the define via something like:
```c
    add r4, r4, #" STR(STREAM1) "\n\
```

Reply:
Thanks, I’m new to inline ASM. I also plan to add a comment explaining why it’s ASM.